### ISSUE TITLE
Build failure on Mac OS with ghc 8.0.1

### ISSUE THREAD

**Issue by gpbaran** [goldfirere/singletons#147]
# Mac OS 10.11.5 with Haskell Platform 8.0.1
Looking forward to learning about the units library, but have been unable to build it under 8.0.1. Running "cabal install units" results in the following:

```
cabal install units
Resolving dependencies...
Downloading singletons-0.10.0...
Configuring singletons-0.10.0...
Building singletons-0.10.0...
Failed to install singletons-0.10.0
Build log ( /Users/gpbaran/.cabal/logs/singletons-0.10.0.log ):
cabal: Entering directory '/var/folders/n9/0737v4gx3hx5r3_l81gt4q7r0000gn/T/cabal-tmp-69455/singletons-0.10.0'
Configuring singletons-0.10.0...
Building singletons-0.10.0...
Preprocessing library singletons-0.10.0...
[ 1 of 19] Compiling Data.Singletons.Void ( src/Data/Singletons/Void.hs, dist/build/Data/Singletons/Void.o )

src/Data/Singletons/Void.hs:8:19: warning: [-Wtrustworthy-safe]
    ‘Data.Singletons.Void’ is marked as Trustworthy but has been inferred as safe!
[ 2 of 19] Compiling Data.Singletons.Types ( src/Data/Singletons/Types.hs, dist/build/Data/Singletons/Types.o )
[ 3 of 19] Compiling Data.Singletons.Util ( src/Data/Singletons/Util.hs, dist/build/Data/Singletons/Util.o )

src/Data/Singletons/Util.hs:22:1: error:
    Failed to load interface for ‘Language.Haskell.TH.Desugar’
    There are files missing in the ‘th-desugar-1.6’ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.

src/Data/Singletons/Util.hs:26:1: error:
    Failed to load interface for ‘Control.Monad.Writer’
    There are files missing in the ‘mtl-2.2.1’ package,
    try running 'ghc-pkg check'.
    Use -v to see a list of the files searched for.
cabal: Leaving directory '/var/folders/n9/0737v4gx3hx5r3_l81gt4q7r0000gn/T/cabal-tmp-69455/singletons-0.10.0'
cabal: Error: some packages failed to install:
singletons-0.10.0 failed during the building phase. The exception was:
ExitFailure 1
units-2.3 depends on singletons-0.10.0 which failed to install.
```

As requested, I ran "ghc-pkg check" and a lot was reported, probably mostly because the build failed. The lines that refer directly to "desugar" are these:

```
There are files missing in the ‘th-desugar-1.6’ package,
There are problems in package th-desugar-1.6:
  Warning: library-dirs: /Users/gpbaran/.cabal/lib/x86_64-osx-ghc-8.0.1/th-desugar-1.6-42XNMpaEzHtFI6XODU1EIz doesn't exist or isn't a directory
  Warning: haddock-interfaces: /Users/gpbaran/.cabal/share/doc/x86_64-osx-ghc-8.0.1/th-desugar-1.6/html/th-desugar.haddock doesn't exist or isn't a file
  Warning: haddock-html: /Users/gpbaran/.cabal/share/doc/x86_64-osx-ghc-8.0.1/th-desugar-1.6/html doesn't exist or isn't a directory
  import-dirs: /Users/gpbaran/.cabal/lib/x86_64-osx-ghc-8.0.1/th-desugar-1.6-42XNMpaEzHtFI6XODU1EIz doesn't exist or isn't a directory
  cannot find any of ["libHSth-desugar-1.6-42XNMpaEzHtFI6XODU1EIz.a","libHSth-desugar-1.6-42XNMpaEzHtFI6XODU1EIz.p_a","libHSth-desugar-1.6-42XNMpaEzHtFI6XODU1EIz-ghc8.0.1.so","libHSth-desugar-1.6-42XNMpaEzHtFI6XODU1EIz-ghc8.0.1.dylib","HSth-desugar-1.6-42XNMpaEzHtFI6XODU1EIz-ghc8.0.1.dll"] on library path
th-desugar-1.6
```

I had loaded the "units" package under 7.10.3 with homebrew ghc, but was enticed by homebrew to try 8.0.1, which broke it, hence this report.

Not too experienced with Haskell, so I had started with Manuel Chakravarty's "Haskell for Mac" which is quite nice up to a point, but restricts you to the stack LTS world only. Hence, this experimentation, since I really want to have dimensioned units for factors in an experimental design package I'm working on. If units were in LTS ...

**Comment by goldfirere** [comment]
I'm not sure why `singletons-0.10` is involved -- that's years old. Only `singletons-2.1` and above are compatible with GHC 8. I've updated the dependency on `base` on Hackage to avoid trying to install `singletons-0.10` on GHC 7.10 or above.

But you've inspired me to get `units` on LTS. I should get it on Stackage nightly today; LTS takes longer.

Bottom line for you: try `cabal install singletons-2.1`. You may get a dependency error, but hopefully that will lead us in the right direction.

Also: did you install Haskell Platform full or minimal?

**Comment by gpbaran** [comment]
Having units in LTS would be great! Thanks.

Also, thanks for helping me out with my confusion. The problems stem mostly from errors I caused through inexperience - but not totally:
- not being sure how broken my system had become from experimenting with homebrew ghc, Haskell for Mac, stack, and Haskell platform, I started over, removing everything Haskell and installing just the full 8.0.1 platform
- but I forgot to "cabal update" which explains the fetching of an ancient version of singletons
- still having problems, discovered I hadn't really cleared everything out by uninstalling, so I nuked ~/.ghc which was the big breakthrough
- building singletons-2.1 then succeeded - great news!
- now units-2.3 mostly builds, but after a while it fails in singletons-0.9.3; units-2.3 looks for versions of singletons above 0.9 but below 2, so it builds 0.9.3.

I'm happy to close this issue because singletons-2.1 builds, but if you like, I can raise another issue for units-2.3 since it won't build because of singletons-0.9.3 as far as I can tell.

Here are the messages from the build of singletons 0.9.3 FYI:

```
Preprocessing library singletons-0.9.3...
[ 1 of 18] Compiling Data.Singletons.Void ( Data/Singletons/Void.hs, dist/build/Data/Singletons/Void.o )
[ 2 of 18] Compiling Data.Singletons.Types ( Data/Singletons/Types.hs, dist/build/Data/Singletons/Types.o )
[ 3 of 18] Compiling Data.Singletons.Util ( Data/Singletons/Util.hs, dist/build/Data/Singletons/Util.o )

Data/Singletons/Util.hs:210:10: error:
    • Could not deduce (Control.Monad.Fail.MonadFail (QWithAux m q))
        arising from the superclasses of an instance declaration
      from the context: (Quasi q, Monoid m)
        bound by the instance declaration
        at Data/Singletons/Util.hs:210:10-52
    • In the instance declaration for ‘Quasi (QWithAux m q)’
```
